### PR TITLE
[Generate] Add bad words list argument to the generate function

### LINKED ISSUE #3061: Bad word list for text generation
# 🚀 Feature request
Add a word of lists that you do not want the model to generate for whatever reason.

## Motivation
When creating a text generation model, especially if you will serve that model publicly, it is desirable to have assurances that a model is physically incapable of outputting certain tokens.

Such tokens would include profanity of all kinds, prejudicial language, or even just out-of-domain vocabulary.

## Your contribution

I am unsure as to your coding style guide, but I will detail how I would implement this below.

Firstly, you cannot simply set the offending word's values to `-Inf` as done here https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_utils.py#L1114 and here https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_utils.py#L1131
when calculating `top_k_top_p_filtering` as it would not capture multi-token words.

As an example, I will use the common Irish exclamation "feck" (see [here ](https://en.wikipedia.org/wiki/Feck) for an off-topic discussion of the word).

'feck' is tokenized into `['fe', 'ck']` in the `gpt2-xl` tokenizer. It would be unreasonable to simply set the logit values of any mention of the token `'fe'` or `'ck'` to -Inf as that would stop the creation of purely harmless words such as 'feasible' (`['fe', 'as', 'ible']`) and 'peck' (`['pe', 'ck']`). Therefore, I suggest a list of tokens that are not allowed at any one instance, and are updated, depending on the previous tokens.
This functionality would mean that a model would be allowed to generate the `'fe'` token, but would not be able to follow it with a `'ck'` token straight after.

I would operationalize this by having:
* a `bad_words_list` which the user has passed into the generate function
* a `possible_bad_words` dict which describes all the possible bad words that you can make with the current token history and how far along they are
* a `prohibited_tokens_list` which would prevent the model from choosing those tokens at a given time.

E.g.
Let's say that we pass the following list to the .generate function:
`bad_words_list = ["feck", "arse", "booze", "up yours"]`
which would be then tokenized and made into ids in the .generator function before the generation starts:
`bad_words_list = [tokenizer.encode(x, add_special_tokens=False) for x in bad_words_list]`

Then, the following function would be run just before the model outputs are obtained from both beam and no beam generators

```
def update_possible_bad_words(previous_token, bad_words_list, possible_bad_words):
    # Start with an empty list of prohibited bad words
    prohibited_tokens_list = []
    unmatched_bad_words= []

    # We cycle through the provided list of bad words, and if a bad word has only one token, then we add it to our prohibited list
    # Else, if a bad word starts with the token of our previous token, then we add it to our list of possible_bad_words
    for bad_word in bad_words_list:
        if len(bad_word) == 1:
            unmatched_bad_words.append(bad_word[0])
        elif previous_token == bad_word[0] and bad_word not in possible_bad_words.keys():
            possible_bad_words[bad_word] = 0

    # We cycle through all our possible bad words
    for bad_word in possible_bad_words.keys():
        bad_word_index = possible_bad_words[bad_word]

        # if the previous token matches the token currently indicated by the stored bad word index, then we increase this index by one.
        if previous_token == bad_word[bad_word_index]:
            new_bad_word_index = bad_word_index +1


            # If the length of the bad word is one greater than the currently stored bad word index, then that means that we need to stop the next token from being generated. Thus, we add it to the prohibited list. We also add this word to the unmatched_bad_words, as we can now consider deleting it from possible bad words as it has been potentially mitigated.
            if len(bad_word) == new_bad_word_index+1:
                prohibited_tokens_list.append(bad_word[-1])
                unmatched_bad_words.append(bad_word)

            # We set the dict value to be this new incremented index
            possible_bad_words[bad_word] = new_bad_word_index 
        else:
           # Else if there is no match (I.e. the pattern of tokens created is different to the bad word) then we can mark this word as unmatched
            unmatched_bad_words.append(bad_word)

    # We cycle through all words that were not matched and we delete them from the possible_bad_words dict, as long as their starting token was not the previous token, in which case it could still be possible bad word.
    for unmatched_bad_word in unmatched_bad_words:
        if previous_token == unmatched_bad_word[0]:
            possible_bad_words[unmatched_bad_word] = 0
        else:
            del possible_bad_words[unmatched_bad_word]

    return prohibited_tokens_list
```
and I would call this function like so:
`prohibited_tokens_list = update_possible_bad_words(input_ids[-1], bad_words_list, possible_bad_words)`
here
https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_utils.py#L827
and here
https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_utils.py#L934
where `possible_bad_words` has been initialized as an empty dict directly before the generation loop.

Finally, we would pass `prohibited_tokens_list` to `top_k_top_p_filtering`
https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_utils.py#L1100
and would simply perform `logits[prohibited_tokens_list] = filter_value` before or after the top p and k filtering in that function.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
The `bad_words_ids` argument allows to insert a list of lists of `input_ids` that cannot be generated, *e.g.* bad words. 

That's a proposed feature request (I think there were actually multiple ones):
#3061

Also adds tests for all language models to verify behavior.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1402,6 +1433,42 @@ def _get_generated_ngrams(hypo_idx):
...
+
+    for prev_input_ids_slice in prev_input_ids:
+        banned_tokens_slice = []
+
+        for banned_token_seq in bad_words_tokens:
+            assert len(banned_token_seq) > 0, "Banned words token sequences {} cannot have an empty list".format(
+                bad_words_tokens
+            )
+
+            if _tokens_match(prev_input_ids_slice.tolist(), banned_token_seq[:-1]) is False:
```
Comment (thomwolf) [inline src/transformers/modeling_utils.py]: Ok seems good for now. At some point in the future, I guess we will want to make generation faster by keeping everything in tensors instead of converting to lists but probably we will need to think deeply about how to make it fast anyway at that time (e.g. switching to a low-level language).

**Review by thomwolf — APPROVED** [review]
Ok great, LGTM!

File: src/transformers/modeling_tf_utils.py [hunk 2]
```diff
@@ -582,6 +586,12 @@ def generate(
             outputs = model.generate(input_ids=input_ids, max_length=50, temperature=0.7, repetition_penalty=1.2)  # generate sequences
             print('Generated: {}'.format(tokenizer.decode(outputs[0], skip_special_tokens=True)))
 
+            tokenizer = AutoTokenizer.from_pretrained('gpt2')   # Initialize tokenizer
+            model = TFAutoModelWithLMHead.from_pretrained('gpt2')    # Download model and configuration from S3 and cache.
+            input_context = 'My cute dog'  # "Legal" is one of the control codes for ctrl
+            bad_words = [tokenizer.encode(bard_word, add_prefix_space=True) for bad_word in ['idiot', 'stupid', 'shut up']]
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_utils.py]: `bard_word` -> `bad_word`

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -582,6 +586,12 @@ def generate(
...
+            input_ids = tokenizer.encode(input_context, return_tensors='tf')  # encode input context
+            outputs = model.generate(input_ids=input_ids, max_length=100, do_sample=True, bad_words_tokens=bad_words_tokens)  # generate sequences without allowing bad_words to be generated
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_utils.py]: `bad_words_tokens` -> `bad_words` ?

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -781,6 +784,12 @@ def generate(
             outputs = model.generate(input_ids=input_ids, max_length=50, temperature=0.7, repetition_penalty=1.2)  # generate sequences
             print('Generated: {}'.format(tokenizer.decode(outputs[0], skip_special_tokens=True)))
 
+            tokenizer = AutoTokenizer.from_pretrained('gpt2')   # Initialize tokenizer
+            model = AutoModelWithLMHead.from_pretrained('gpt2')    # Download model and configuration from S3 and cache.
+            input_context = 'My cute dog'  # "Legal" is one of the control codes for ctrl
+            bad_words = [tokenizer.encode(bard_word, add_prefix_space=True) for bad_word in ['idiot', 'stupid', 'shut up']]
+            input_ids = tokenizer.encode(input_context, return_tensors='pt')  # encode input context
+            outputs = model.generate(input_ids=input_ids, max_length=100, do_sample=True, bad_words_tokens=bad_words_tokens)  # generate sequences without allowing bad_words to be generated
```
Comment (LysandreJik) [inline src/transformers/modeling_utils.py]: same as above

**Review by LysandreJik — APPROVED** [review]
Nice work! Left a few comments related to the documentation. When these are fixed, feel free to merge @patrickvonplaten!

**Comment by patrickvonplaten** [comment]
Sadly, TF tensorflow test seems flaky see: https://github.com/huggingface/transformers/commit/b38d552a92a0a201c005afae0e1b861ae6de9ce0

Might need to revert the commit.

**Comment by Alvant** [comment]
@patrickvonplaten Could I please ask you a question about `bad_words_ids`? I am wondering, why is it made as `bad_words_ids` and not as `bad_words_list`? That is, why the parameter is devised to be a list of lists of token ids and not just a list of words as strings? Was it made so because of some implementation constraints? or is there some advantages for the user? Well, in other words, why is `bad_words_ids` better than `bad_words_list`? :slightly_smiling_face: 

The reason I am asking this is that I want to implement similar feature in another repository, and so would be very glad to hear you thoughts on the issue as the author who implemented this in transformers)

P.S. Sorry for intruding this PR with questions after several years since it was merged :sweat_smile:
